core/src/main/java/org/elasticsearch/plugins/PluginsService.java

```diff
@@ -290,10 +291,13 @@ public Settings updatedSettings() {
         return builders;
     }
 
-    public Collection<Class<? extends LifecycleComponent>> nodeServices() {
-        List<Class<? extends LifecycleComponent>> services = new ArrayList<>();
+    /** Returns all {@link LifecycleComponent} instances, whether created by guice or plugins. */
+    public Collection<LifecycleComponent> nodeServices(Injector injector) {
+        List<LifecycleComponent> services = new ArrayList<>();
         for (Tuple<PluginInfo, Plugin> plugin : plugins) {
-            services.addAll(plugin.v2().nodeServices());
+            services.addAll(plugin.v2().getGuiceServices().stream().map(injector::getInstance).collect(Collectors.toList()));
+            services.addAll(plugin.v2().getComponents().stream().filter(p -> p instanceof LifecycleComponent)
```

Comment:
This presumes that calling `getComponents` is without side effects but I kind of expect it to return different instances each time it is called.